### PR TITLE
email testing fixture

### PR DESCRIPTION
**Pull request by ice-cronus**
Function to read validation code from delivered email
Generator of email addresses using https://www.1secmail.com/

### CODE REVIEW

File: email/fixture/fixture.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+func GetEmailConfirmationCode(ctx context.Context, tb testing.TB, toEmail string) string {
+	tb.Helper()
+	emailSplitted := strings.Split(toEmail, "@")
+	assert.Len(tb, emailSplitted, 2) //nolint:gomnd // Not a magic.
+	login, server := emailSplitted[0], emailSplitted[1]
+	var messageListData []*struct {
+		ID uint64 `json:"id"`
+	}
+	url := fmt.Sprintf("https://www.1secmail.com/api/v1/?action=getMessages&login=%v&domain=%v", login, server)
+	for ctx.Err() == nil && len(messageListData) == 0 {
```
[the remark this answers is not in the document]
Comment (ice-cronus) [reply email/fixture/fixture.go:31]: I used the same in the awaiting sms delivery loop in eskimo's fixture, and it died properly when timeout expired (with deadline exceeded), I will recheck.

[the remark this answers is not in the document]
Comment (ice-myles) [reply email/fixture/fixture.go:31]: Okay, nice if so.

**Review by ice-myles — DISMISSED** [review]

File: email/fixture/fixture.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	messageInfoBody, messageInfoStatus := do1SecMailRequest(ctx, tb, url)
+	assert.Equal(tb, 200, messageInfoStatus) //nolint:gomnd // Not a magic.
+	var messageInfo struct {
+		From    string `json:"from"`
+		Subject string `json:"subject"`
+		Body    string `json:"body"`
+	}
+	require.NoError(tb, json.Unmarshal(messageInfoBody, &messageInfo))
+	verificationCode := parseVerificationMessage(messageInfo.Body)
+	assert.Len(tb, verificationCode, 6) //nolint:gomnd // 6-digit long code
```
Comment (ice-ares) [inline email/fixture/fixture.go:50]: hell no. remove it, replace it with comparing the actual code and dont return it

File: email/fixture/fixture.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/http2"
+)
+
+func GetEmailConfirmationCode(ctx context.Context, tb testing.TB, toEmail string) string {
```
Comment (ice-ares) [inline email/fixture/fixture.go:22]: Assert instead of Get, and add the expected code as arg

File: email/fixture/fixture.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	assert.Equal(tb, 200, status) //nolint:gomnd // Not a magic.
+	var emails []string
+	require.NoError(tb, json.Unmarshal(body, &emails))
+	assert.Len(tb, emails, 1)
+
+	return emails[0]
+}
+
+func parseVerificationMessage(msg string) string {
+	r := regexp.MustCompile(`^<strong>\d+`)
```
Comment (ice-ares) [inline email/fixture/fixture.go:67]: nope, thats implementation specific, provide a func for the user to implement the parsing

File: email/fixture/fixture.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/stretchr/testify/require"
+	"golang.org/x/net/http2"
+)
+
+func GetEmailConfirmationCode(ctx context.Context, tb testing.TB, toEmail string) string {
+	tb.Helper()
+	emailSplitted := strings.Split(toEmail, "@")
+	assert.Len(tb, emailSplitted, 2) //nolint:gomnd // Not a magic.
+	login, server := emailSplitted[0], emailSplitted[1]
+	var messageListData []*struct {
```
Comment (ice-ares) [inline email/fixture/fixture.go:27]: i would extract into separate method: getMailbox

File: email/fixture/fixture.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+		require.NoError(tb, json.Unmarshal(messageListBody, &messageListData))
+		time.Sleep(100 * time.Millisecond) //nolint:gomnd // Not a magic.
+	}
+	if len(messageListData) == 0 {
+		require.Fail(tb, "Expected email message, but it was not delivered")
+	}
+	url = fmt.Sprintf("https://www.1secmail.com/api/v1/?action=readMessage&login=%v&domain=%v&id=%v", login, server, messageListData[0].ID)
+	messageInfoBody, messageInfoStatus := do1SecMailRequest(ctx, tb, url)
+	assert.Equal(tb, 200, messageInfoStatus) //nolint:gomnd // Not a magic.
+	var messageInfo struct {
```
Comment (ice-ares) [inline email/fixture/fixture.go:43]: extract into new method: verifyEmailMessage

**Review by ice-ares — CHANGES_REQUESTED** [review]

File: email/fixture/fixture.go (hunk 3 above)
Comment (ice-cronus) [reply email/fixture/fixture.go]: We still need Get for cmd tests cuz we have no access to database there to get actual code (I removed that strange fixture to find DB connector as we can get code from email now), but we need the code to finish the flow. So I added both functions, that's better than getting DB connector in cmd part.

Comment (ice-ares) [reply email/fixture/fixture.go]: U dont need to get it from db, u get it from the body

Comment (ice-cronus) [reply email/fixture/fixture.go]: I mean if we'll have only Assert...(expectedCode string), I'll need to get expected one somewhere, in internal tests I can get it from DB, but for cmd we still need Get..() to get it from the body

File: email/fixture/fixture.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	"golang.org/x/net/http2"
+)
+
+func AssertEmailConfirmationCode(ctx context.Context, tb testing.TB, toEmail, expectedCode string, parseMsg func(msg string) string) {
+	tb.Helper()
+	verificationCode := GetEmailConfirmationCode(ctx, tb, toEmail, parseMsg)
+	assert.Equal(tb, expectedCode, verificationCode)
+}
+
+func GetEmailConfirmationCode(ctx context.Context, tb testing.TB, toEmail string, parseMsg func(msg string) string) string {
```
Comment (ice-ares) [inline email/fixture/fixture.go:27]: U dont need this

File: email/fixture/fixture.go (hunk 3 above)
Comment (ice-ares) [reply email/fixture/fixture.go]: You get it to do what?

Comment (ice-cronus) [reply email/fixture/fixture.go]: to assert the expected one (db's/somewhere's else value) with the actual, from email

File: email/fixture/fixture.go (hunk 7 above)
Comment (ice-ares) [reply email/fixture/fixture.go:27]: Ph, now i get it

File: email/fixture/fixture.go (hunk 3 above)
Comment (ice-cronus) [reply email/fixture/fixture.go]: And also I need them to confirm an email with correct code

Comment (ice-ares) [reply email/fixture/fixture.go]: Nope, u need it to pass it as arg to the validate email endpoint

**Review by ice-ares — APPROVED** [review]

**Review by ice-laius — APPROVED** [review]

**Pull request merged**